sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -0,0 +1,465 @@
+      private final PreparedStatementSetter<T> preparedStatementSetter;
+      private long batchSize;
+
+      private Connection connection;
+      private List<T> batch;
+
+      public JdbcWriter(DataSource dataSource, String query, String username, String password,
+                        PreparedStatementSetter<T> preparedStatementSetter, long batchSize) {
+        this.dataSource = dataSource;
+        this.query = query;
+        this.username = username;
+        this.password = password;
+        this.preparedStatementSetter = preparedStatementSetter;
+        this.batchSize = batchSize;
+      }
+
+      public JdbcWriter<T> withDataSource(DataSource dataSource) {
+        return new JdbcWriter<>(dataSource, query, username, password, preparedStatementSetter,
+            batchSize);
+      }
+
+      public JdbcWriter<T> withQuery(String query) {
+        return new JdbcWriter<>(dataSource, query, username, password, preparedStatementSetter,
+            batchSize);
+      }
+
+      public JdbcWriter<T> withUsername(String username) {
+        return new JdbcWriter<>(dataSource, query, username, password, preparedStatementSetter,
+            batchSize);
+      }
+
+      public JdbcWriter<T> withPassword(String password) {
+        return new JdbcWriter<>(dataSource, query, username, password, preparedStatementSetter,
+            batchSize);
+      }
+
+      public JdbcWriter<T> withPreparedStatementSetter(
+          PreparedStatementSetter<T> preparedStatementSetter) {
+        return new JdbcWriter<>(dataSource, query, username, password, preparedStatementSetter,
+            batchSize);
+      }
+
+      public JdbcWriter<T> withBatchSize(long batchSize) {
+        return new JdbcWriter<>(dataSource, query, username, password, preparedStatementSetter,
+            batchSize);
+      }
+
+      public void validate() {
+        Preconditions.checkNotNull(dataSource, "dataSource");
+        Preconditions.checkNotNull(query, "query");
+        Preconditions.checkNotNull(preparedStatementSetter, "preparedStatementSetter");
+      }
+
+      @Setup
+      public void connectToDatabase() throws Exception {
+        if (username != null) {
+          connection = dataSource.getConnection(username, password);
+        } else {
+          connection = dataSource.getConnection();
+        }
+        connection.setAutoCommit(true);
+      }
+
+      @StartBundle
+      public void startBundle(Context context) {
+        batch = new ArrayList<>();
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) throws Exception {
+        T record = context.element();
+
+        batch.add(record);
+        if (batch.size() >= batchSize) {
+          finishBundle(context);
+        }
+      }
+
+      @FinishBundle
+      public void finishBundle(Context context) throws Exception {
+        for (T record : batch) {
+          try {
+            PreparedStatement statement = connection.prepareStatement(query);
```

Comment:
This code is preparing the statement once per element. It's going to be very, very slow.
You need to prepare the statement as rarely as possible - ideally in Setup; worse, in startBundle; even worse, once per flushed batch.
Then you also need to clear the parameters of the statement before calling setParameters.